File: kotlin/src/main/kotlin/com/cyaim/im/client/Protocol.kt
@file:OptIn(ExperimentalSerializationApi::class)

package com.cyaim.im.client

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNames
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject

/*
 * Wire protocol types. These mirror docs/SPEC-02-protocol.md and sdk/CONTRACT.md §2, and are the
 * only place the shape of a frame is described on the client side. Payload models live in
 * Models.kt, request objects in Requests.kt, enumerations in Enums.kt.
 */

/**
 * The exact codec the SDK uses on the socket. It is public because your own payloads travel inside
 * [ImMessage.content] and [ImBody.data] as raw [JsonElement]s — decode them with this instance and
 * they get the same leniency the SDK gives itself.
 *
 * The four settings are CONTRACT.md §2's JSON policy, one for one:
 *
 * - `ignoreUnknownKeys` — the gateway is deployed far more often than the app on a user's phone,
 *   so every field the server adds after a release has to be invisible to clients built before it.
 *   A strict decoder turns "the server added a field" into "the app stopped receiving messages".
 * - `explicitNulls = false` — nulls are omitted when writing; an explicit null read into a
 *   non-nullable field falls back to its default rather than throwing.
 * - `isLenient` — `NumberHandling.AllowReadingFromString` on the server means a 64-bit `seq` can
 *   arrive as `"1234"`. Lenient parsing is what stops that from being an exception.
 * - `coerceInputValues` — the same forgiveness for a null where a default will do.
 */
public val ImJson: Json = Json {
    ignoreUnknownKeys = true
    isLenient = true
    explicitNulls = false
    coerceInputValues = true
    encodeDefaults = true
}

/** A request the client sends. `id` is mandatory: it is how a reply finds its caller. */
@Serializable
public data class ImRequest(
    public val id: String,
    public val target: String,
    public val body: JsonElement = JsonObject(emptyMap()),
)

/**
 * A frame the server sends. Replies and server-initiated pushes are structurally identical, which
 * is deliberate: one decoder handles both, and a push can be correlated exactly like a reply.
 *
 * Every field carries a PascalCase alternate. SPEC-02 §1.2 writes the envelope as
 * `Id`/`Target`/`Status`/`Body` while the gateway's JSON options emit camelCase, and a Kotlin
 * decoder — unlike a structurally typed one — quietly produces an all-defaults object when it
 * recognises none of the keys. That failure mode is a client which connects, receives frames and
 * delivers nothing, so we accept both spellings rather than bet on one.
 */
@Serializable
public data class ImFrame(
    @JsonNames("Id") public val id: String = "",
    @JsonNames("Target") public val target: String = "",
    /** Transport-level outcome: 0 routed, 1 endpoint threw, 2 endpoint not found. */
    @JsonNames("Status") public val status: Int = 0,
    @JsonNames("Msg") public val msg: String? = null,
    @JsonNames("RequestTime") public val requestTime: Long? = null,
    @JsonNames("CompleteTime") public val completeTime: Long? = null,
    @JsonNames("Body") public val body: ImBody? = null,
)

/**
 * Business-level result, nested inside the transport frame.
 *
 * The nesting is deliberate on the server side too: [ImFrame.status] is the transport outcome
 * (did the frame reach an endpoint at all) and [code] is the business outcome. Collapsing them
 * would lose the difference between "the gateway could not route this" and "the endpoint ran and
 * said no".
 */
@Serializable
public data class ImBody(
    @JsonNames("Code") public val code: Int = 0,
    @JsonNames("Message") public val message: String? = null,
    @JsonNames("TraceId") public val traceId: String? = null,
    @JsonNames("ServerTime") public val serverTime: Long = 0,
    @JsonNames("Data") public val data: JsonElement? = null,
)

/**
 * Business error codes (SPEC-02 §4, CONTRACT.md §7).
 *
 * Constants rather than an enum, on purpose: codes arrive off the wire from a server that is
 * newer than this artifact, and an enum would have to either fail to decode or fold the unknown
 * value into an `Unknown` member — throwing away the exact number the support ticket needs.
 *
 * The list is the server's `ImErrorCode` in full rather than the handful this SDK happens to
 * raise. A tenant branching on `1703 StorageQuotaExceeded` should not have to hard-code the
 * number because the SDK only bothered to name the codes it produces itself.
 */
public object ImErrorCode {
    public const val Ok: Int = 0

    // ---- 1000: transport and generic ----
    public const val InternalError: Int = 1000
    public const val InvalidArgument: Int = 1001
    public const val NotFound: Int = 1002

    /** Per-tenant API quota, or an endpoint's own limiter. Back off and surface it; see §7.4. */
    public const val RateLimited: Int = 1003
    public const val Timeout: Int = 1004

    /**
     * Also what a call made while the socket is down raises, client-side, without leaving the
     * device. CONTRACT.md §7.2 fixes this code for that case rather than inventing a local one.
     */
    public const val ServiceUnavailable: Int = 1005
    public const val Conflict: Int = 1006
    public const val PayloadTooLarge: Int = 1007

    /**
     * The deployment does not have this endpoint — an SDK newer than a private-deployment server.
     * Deliberately not [NotFound]: 1002 means "your group does not exist".
     */
    public const val UnsupportedOperation: Int = 1008

    /** Kept for source compatibility. [ServiceUnavailable] is the contract name for 1005. */
    @Deprecated("Renamed to ServiceUnavailable to match the server enum", ReplaceWith("ImErrorCode.ServiceUnavailable"))
    public const val NotConnected: Int = 1005

    // ---- 1100: authentication ----
    public const val Unauthorized: Int = 1100
    public const val TokenExpired: Int = 1101
    public const val TokenInvalid: Int = 1102
    public const val Forbidden: Int = 1103
    public const val UserBanned: Int = 1104
    public const val SignatureInvalid: Int = 1105
    public const val ReplayDetected: Int = 1106

    /** Arrives both as a `conn.kick` push and as a close reason. Never reconnect into it. */
    public const val KickedByOtherDevice: Int = 1107

    // ---- 1200: tenant and plan ----
    public const val AppNotFound: Int = 1200
    public const val AppDisabled: Int = 1201
    public const val QuotaExceeded: Int = 1202

    /**
     * The tenant has the feature switched off, or the module is not deployed.
     * **Never latch this locally** — a tenant can flip the flag at runtime and a client that
     * remembers "typing is off" stays broken until the app restarts.
     */
    public const val FeatureNotEnabled: Int = 1203
    public const val PlanExpired: Int = 1204
    public const val ConcurrencyLimitExceeded: Int = 1205

    // ---- 1300: users and relationships ----
    public const val UserNotFound: Int = 1300
    public const val UserAlreadyExists: Int = 1301
    public const val NotFriend: Int = 1302
    public const val BlockedByPeer: Int = 1303
    public const val BlockedPeer: Int = 1304
    public const val FriendRequestNotFound: Int = 1305
    public const val FriendLimitExceeded: Int = 1306
    public const val CannotAddSelf: Int = 1307

    // ---- 1400: messages ----
    public const val MessageNotFound: Int = 1400
    public const val MessageTooLong: Int = 1401
    public const val ModerationRejected: Int = 1402
    public const val RecallWindowExpired: Int = 1403
    public const val RecallForbidden: Int = 1404
    public const val EditWindowExpired: Int = 1405
    public const val DuplicateClientMessageId: Int = 1406
    public const val ConversationNotFound: Int = 1407
    public const val SenderMuted: Int = 1408
    public const val UnsupportedContentType: Int = 1409
    public const val ReceiptDisabled: Int = 1410

    // ---- 1500: groups ----
    public const val GroupNotFound: Int = 1500
    public const val GroupDismissed: Int = 1501
    public const val GroupFull: Int = 1502
    public const val NotGroupMember: Int = 1503
    public const val NoGroupPermission: Int = 1504
    public const val GroupMuted: Int = 1505
    public const val MemberMuted: Int = 1506
    public const val AlreadyGroupMember: Int = 1507
    public const val JoinNeedsApproval: Int = 1508
    public const val JoinForbidden: Int = 1509
    public const val InviteForbidden: Int = 1510
    public const val CannotOperateOwner: Int = 1511
    public const val ApplicationNotFound: Int = 1512

    // ---- 1600: chat rooms ----
    public const val RoomNotFound: Int = 1600
    public const val RoomFull: Int = 1601
    public const val NotInRoom: Int = 1602
    public const val RoomMuted: Int = 1603

    // ---- 1700: media ----
    public const val UploadFailed: Int = 1700
    public const val FileTypeNotAllowed: Int = 1701
    public const val FileTooLarge: Int = 1702
    public const val StorageQuotaExceeded: Int = 1703

    // ---- 2400: push delivery ----

    /**
     * A click named a delivery row this tenant does not have: it aged out after seven days, or the
     * notification was never sent by this platform.
     *
     * The one code of the 2400 band a device ever sees — the rest of it belongs to the console
     * plane, where a signed-in member reads their own push records. [PushApi.clicked] does not
     * raise it; see there for why a click count that is one short is not an application's problem.
     */
    public const val PushDeliveryNotFound: Int = 2401

    /**
     * Codes worth trying again, unchanged, later. Computed from the code alone so that all five
     * SDKs agree (CONTRACT.md §7.3) — note that nothing outside this set is retryable, including
     * every 1100 and 1200 code: retrying those produces the same answer.
     */
    public fun isRetryable(code: Int): Boolean =
        code == InternalError || code == RateLimited || code == Timeout || code == ServiceUnavailable

    /**
     * Codes that mean "the token, not the request, is what was rejected".
     *
     * Narrower than the whole 1100 band on purpose: [Forbidden] and [UserBanned] also live there
     * and a fresh token fixes neither.
     */
    public fun requiresReauth(code: Int): Boolean =
        code == Unauthorized || code == TokenExpired || code == TokenInvalid
}

/** Server-initiated event names (SPEC-02 §2.7). */
public object PushTarget {
    public const val Message: String = "evt.message"
    public const val MessageUpdate: String = "evt.messageUpdate"
    public const val ConversationUpdate: String = "evt.conversationUpdate"
    public const val Read: String = "evt.read"
    public const val Typing: String = "evt.typing"
    public const val Presence: String = "evt.presence"
    public const val Friend: String = "evt.friend"
    public const val Group: String = "evt.group"
    public const val System: String = "evt.system"
    public const val Stream: String = "evt.stream"
    public const val Call: String = "evt.call"
    public const val Desk: String = "evt.desk"
    public const val Kick: String = "conn.kick"
}

/**
 * Maps an open enum to its wire integer and back. Unlike a generated enum serializer it cannot
 * fail: every integer is a legal value, and an unrecognised one keeps its number (see Enums.kt).
 */
public class IntCodeSerializer<T>(
    serialName: String,
    private val encode: (T) -> Int,
    private val decode: (Int) -> T,
) : KSerializer<T> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor(serialName, PrimitiveKind.INT)
    override fun serialize(encoder: Encoder, value: T): Unit = encoder.encodeInt(encode(value))
    override fun deserialize(decoder: Decoder): T = decode(decoder.decodeInt())
}

/**
 * Every non-zero business code becomes one of these, so callers can `try`/`catch` instead of
 * inspecting each result.
 *
 * Named `Exception` rather than `Error` (the TypeScript SDK's `ImError`) because on the JVM
 * `Error` means "the runtime is broken, do not catch this", and every one of these is catchable
 * and usually recoverable.
 */
public class ImException(
    /** A value from [ImErrorCode]. Branch on the constants, never on the message text. */
    public val code: Int,
    message: String,
    /** Server-side correlation id. Quote it in a support ticket and the server logs line up. */
    public val traceId: String? = null,
    /** The endpoint that failed, e.g. `msg.send`. */
    public val target: String? = null,
    cause: Throwable? = null,
) : Exception(message, cause) {

    /**
     * True when retrying the identical call could succeed. See [ImErrorCode.isRetryable].
     *
     * The SDK never acts on this by itself: it retries the *connection* and its own gap repair,
     * and nothing else. An SDK that silently re-sends on `1003` hides rate limiting from the UI
     * that has to explain it, and one that silently retries a send turns a visible failure into
     * an invisible delay.
     */
    public val isRetryable: Boolean get() = ImErrorCode.isRetryable(code)

    /** True when a fresh token would fix this. See [ImErrorCode.requiresReauth]. */
    public val requiresReauth: Boolean get() = ImErrorCode.requiresReauth(code)

    /** The whole 1100–1199 band. [requiresReauth] is the narrower question and usually the one you want. */
    public val isAuthFailure: Boolean get() = code in 1100..1199

    override fun toString(): String = buildString {
        append("ImException(code=").append(code)
        target?.let { append(", target=").append(it) }
        traceId?.let { append(", traceId=").append(it) }
        append("): ").append(message)
    }
}

/** Lifecycle of the socket, published as a `StateFlow` so UI can bind to it directly. */
public enum class ConnectionState {
    Idle,
    Connecting,
    Open,
    Reconnecting,
    Closed,
}

/**
 * Why the server hung up on us.
 *
 * The server closes with reason `im-kick:{Reason}`. Terminal reasons are the ones where coming
 * back would fail identically, forever — retrying them is how you ship a client that hammers the
 * gateway with a revoked token until the battery is flat.
 */
public enum class KickReason {
    /** Another device of the same user took the slot the multi-login policy allows. */
    MultiLoginPolicy,

    /** The token was invalidated server-side: logout elsewhere, key rotation. */
    TokenRevoked,

    /** The account is banned. */
    UserBanned,

    /** The tenant application was disabled — usually billing. */
    AppDisabled,

    /** A console operator forced this session off. */
    AdminKick,

    /** Recoverable: ask the host app for a fresh token, then come back. */
    TokenExpired,

    /** A reason this SDK version does not know. Treated as recoverable; see [isTerminal]. */
    Unknown,
    ;

    /** Terminal reasons must not reconnect. */
    public val isTerminal: Boolean
        get() = this == MultiLoginPolicy ||
            this == TokenRevoked ||
            this == UserBanned ||
            this == AppDisabled ||
            this == AdminKick

    public companion object {
        /** Prefix of the WebSocket close reason that marks a kick rather than a network failure. */
        public const val ClosePrefix: String = "im-kick:"

        public fun fromWire(name: String): KickReason =
            entries.firstOrNull { it.name.equals(name, ignoreCase = true) } ?: Unknown
    }
}

/** A kick, as delivered on [ImClient.kicks]. */
public data class KickEvent(
    public val reason: KickReason,
    /** The raw text after `im-kick:`, kept so an unrecognised reason is still reportable. */
    public val raw: String,
)

/**
 * Collapses "absent" and "JSON null" into an empty object.
 *
 * Endpoints that return nothing send no `data` at all, and decoding `Unit` — or anything else —
 * from an empty object succeeds where decoding from null throws. One coercion here removes a
 * special case from every call site.
 */
internal fun JsonElement?.orEmptyObject(): JsonElement =
    if (this == null || this is JsonNull) JsonObject(emptyMap()) else this

File: swift/Sources/CyaimIM/Protocol.swift
import Foundation

// The wire protocol. This file mirrors docs/SPEC-02-protocol.md and is the only place the shape of
// a frame is described on the client side.
//
// The identifier-like types below (error codes, push targets, kick reasons, platforms, content
// types) are `RawRepresentable` structs rather than enums on purpose: the server is a moving
// product and will add a content type, an event or an error code long before a given app is
// rebuilt. An unknown raw value has to decode to *itself*, not throw — a shipped client that
// fails to parse a value added six months later is worse than one that ignores it.

// MARK: - Error codes

/// Business result codes carried inside the response body. Stable API surface: values never move.
public struct ImErrorCode: RawRepresentable, Hashable, Sendable, Codable, CustomStringConvertible {
    public let rawValue: Int

    public init(rawValue: Int) { self.rawValue = rawValue }
    public init(_ rawValue: Int) { self.rawValue = rawValue }

    public init(from decoder: any Decoder) throws {
        self.init(rawValue: try decoder.singleValueContainer().decode(Int.self))
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }

    public var description: String { String(rawValue) }

    public static let ok = ImErrorCode(0)

    // 1000-1099 generic
    public static let internalError = ImErrorCode(1000)
    public static let invalidArgument = ImErrorCode(1001)
    public static let notFound = ImErrorCode(1002)
    public static let rateLimited = ImErrorCode(1003)
    public static let timeout = ImErrorCode(1004)
    public static let serviceUnavailable = ImErrorCode(1005)

    /// Client-side only: the socket was down, so the request never left the device. Shares 1005
    /// with `serviceUnavailable` deliberately — to a caller deciding whether to retry, "the server
    /// is not there" and "the pipe to it is not there" are the same answer, and the other SDKs
    /// report this value under their own name for it.
    public static let notConnected = ImErrorCode(1005)
    public static let conflict = ImErrorCode(1006)
    public static let payloadTooLarge = ImErrorCode(1007)
    public static let unsupportedOperation = ImErrorCode(1008)

    // 1100-1199 auth
    public static let unauthorized = ImErrorCode(1100)
    public static let tokenExpired = ImErrorCode(1101)
    public static let tokenInvalid = ImErrorCode(1102)
    public static let forbidden = ImErrorCode(1103)
    public static let userBanned = ImErrorCode(1104)
    public static let signatureInvalid = ImErrorCode(1105)
    public static let replayDetected = ImErrorCode(1106)
    public static let kickedByOtherDevice = ImErrorCode(1107)

    /// The page's origin is not on the app's web allowlist. A tenant sets that list in the console; it does not vary by user, so retrying or re-authenticating will not help and the SDK must not.
    ///
    /// 页面来源不在该应用的 Web 安全域名表里。这张表由租户在控制台设置、与用户无关——重试或重新登录都没有用，SDK 也不该那么做。
    public static let originNotAllowed = ImErrorCode(1109)

    // 1200-1299 tenant & quota
    public static let appNotFound = ImErrorCode(1200)
    public static let appDisabled = ImErrorCode(1201)

    /// Tenant billing. Never retry, and never bury it in a generic failure: the integrator
    /// debugging this needs the code.
    public static let quotaExceeded = ImErrorCode(1202)

    /// The tenant has this feature switched off, or the module is not deployed.
    ///
    /// **Do not latch it.** A tenant can flip a flag at runtime, and a client that remembers
    /// "typing is off" stays broken until the app restarts. Report it every time and keep calling.
    public static let featureNotEnabled = ImErrorCode(1203)

    public static let planExpired = ImErrorCode(1204)
    public static let concurrencyLimitExceeded = ImErrorCode(1205)

    // 1300-1399 user & relationship
    public static let userNotFound = ImErrorCode(1300)
    public static let userAlreadyExists = ImErrorCode(1301)
    public static let notFriend = ImErrorCode(1302)
    public static let blockedByPeer = ImErrorCode(1303)
    public static let blockedPeer = ImErrorCode(1304)
    public static let friendRequestNotFound = ImErrorCode(1305)
    public static let friendLimitExceeded = ImErrorCode(1306)
    public static let cannotAddSelf = ImErrorCode(1307)

    // 1400-1499 message
    public static let messageNotFound = ImErrorCode(1400)
    public static let messageTooLong = ImErrorCode(1401)
    public static let moderationRejected = ImErrorCode(1402)
    public static let recallWindowExpired = ImErrorCode(1403)
    public static let recallForbidden = ImErrorCode(1404)
    public static let editWindowExpired = ImErrorCode(1405)
    public static let duplicateClientMessageId = ImErrorCode(1406)
    public static let conversationNotFound = ImErrorCode(1407)
    public static let senderMuted = ImErrorCode(1408)
    public static let unsupportedContentType = ImErrorCode(1409)
    public static let receiptDisabled = ImErrorCode(1410)

    // 1500-1599 group
    public static let groupNotFound = ImErrorCode(1500)
    public static let groupDismissed = ImErrorCode(1501)
    public static let groupFull = ImErrorCode(1502)
    public static let notGroupMember = ImErrorCode(1503)
    public static let noGroupPermission = ImErrorCode(1504)
    public static let groupMuted = ImErrorCode(1505)
    public static let memberMuted = ImErrorCode(1506)
    public static let alreadyGroupMember = ImErrorCode(1507)

    /// Not a failure to render as one: `group.join` lodged an application and it is pending.
    public static let joinNeedsApproval = ImErrorCode(1508)

    public static let joinForbidden = ImErrorCode(1509)
    public static let inviteForbidden = ImErrorCode(1510)
    public static let cannotOperateOwner = ImErrorCode(1511)
    public static let applicationNotFound = ImErrorCode(1512)

    // 1600-1699 chat room
    public static let roomNotFound = ImErrorCode(1600)
    public static let roomFull = ImErrorCode(1601)
    public static let notInRoom = ImErrorCode(1602)
    public static let roomMuted = ImErrorCode(1603)

    // 1700-1799 media & storage
    public static let uploadFailed = ImErrorCode(1700)
    public static let fileTypeNotAllowed = ImErrorCode(1701)
    public static let fileTooLarge = ImErrorCode(1702)
    public static let storageQuotaExceeded = ImErrorCode(1703)

    /// No delivery record matches. The row is kept seven days, or the notification did not come from this platform at all. `clicked` swallows it: a click count one short is not an application's problem, and there is nothing a user could do about it.
    ///
    /// 没有匹配的投递记录：记录只保留七天，或者这条通知根本不是本平台发的。
    public static let pushDeliveryNotFound = ImErrorCode(2401)
}

/// Every failure the SDK reports, carrying the code and the server's trace id.
///
/// `traceId` is much of the point of this type. When a tenant opens a support ticket, the trace id
/// is what turns "sending fails sometimes" into one request on one node in one log, so it is
/// surfaced all the way to the caller instead of being logged and dropped.
public struct ImError: Error, Sendable, Hashable, CustomStringConvertible, LocalizedError {
    /// Business code from the response body, or a client-side code for transport failures.
    public let code: ImErrorCode

    public let message: String

    /// Server-side correlation id. Quote it in a support ticket.
    public let traceId: String?

    /// The endpoint that failed, e.g. `msg.send`.
    public let target: String?

    public init(code: ImErrorCode, message: String, traceId: String? = nil, target: String? = nil) {
        self.code = code
        self.message = message
        self.traceId = traceId
        self.target = target
    }

    public var description: String {
        var text = "ImError(\(code.rawValue)): \(message)"
        if let target { text += " [\(target)]" }
        if let traceId { text += " trace=\(traceId)" }
        return text
    }

    public var errorDescription: String? { description }

    /// True when retrying with the *same* payload is safe and sensible — the send may never have
    /// landed, and `clientMsgId` makes a repeat idempotent either way.
    ///
    /// Exactly `1000` InternalError, `1003` RateLimited, `1004` Timeout, `1005`
    /// ServiceUnavailable, and nothing else. Everything outside that list is terminal: retrying it
    /// produces the same answer.
    ///
    /// The SDK **never** retries a business call on your behalf. It retries the *connection*, and
    /// its own `conn.sync` / `msg.sync` repair; everything else surfaces with this flag set and the
    /// application decides. Silently re-sending on `1003` hides rate limiting from the UI that has
    /// to explain it, and silently retrying a send turns a visible failure into an invisible delay.
    public var isRetryable: Bool {
        switch code {
        case .internalError, .rateLimited, .timeout, .serviceUnavailable: return true
        default: return false
        }
    }

    /// True when the host app should mint a new token before trying again.
    ///
    /// `1100` Unauthorized, `1101` TokenExpired, `1102` TokenInvalid. Note that the SDK already
    /// handles `1101` on a live socket by calling `conn.reauth` with a token from
    /// ``ImClientOptions/tokenProvider`` and retrying once — this flag is what is left when that
    /// did not work, or when no provider was configured.
    public var requiresReauth: Bool {
        switch code {
        case .unauthorized, .tokenExpired, .tokenInvalid: return true
        default: return false
        }
    }

    /// The contract's name for this flag is ``requiresReauth``; this spelling shipped first and
    /// stays for source compatibility. Deprecated for 2.0.
    public var isAuthFailure: Bool { requiresReauth }
}

// MARK: - Push targets

/// Server-initiated event names.
public struct PushTarget: RawRepresentable, Hashable, Sendable, Codable, CustomStringConvertible {
    public let rawValue: String

    public init(rawValue: String) { self.rawValue = rawValue }
    public init(_ rawValue: String) { self.rawValue = rawValue }

    public init(from decoder: any Decoder) throws {
        self.init(rawValue: try decoder.singleValueContainer().decode(String.self))
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }

    public var description: String { rawValue }

    public static let message = PushTarget("evt.message")
    public static let messageUpdate = PushTarget("evt.messageUpdate")
    public static let conversationUpdate = PushTarget("evt.conversationUpdate")
    public static let read = PushTarget("evt.read")
    public static let typing = PushTarget("evt.typing")
    public static let presence = PushTarget("evt.presence")
    public static let friend = PushTarget("evt.friend")
    public static let group = PushTarget("evt.group")
    public static let system = PushTarget("evt.system")
    public static let stream = PushTarget("evt.stream")
    public static let call = PushTarget("evt.call")
    public static let desk = PushTarget("evt.desk")

    /// Not an `evt.*` event: the server pushes this immediately before closing the socket.
    public static let kick = PushTarget("conn.kick")
}

// MARK: - Kick reasons

/// Why the server closed a connection, parsed from the WebSocket close reason `im-kick:{Reason}`.
///
/// This is the only signal that separates "you were kicked" from "the network died", and the two
/// need opposite responses: one must never reconnect, the other must.
public struct KickReason: RawRepresentable, Hashable, Sendable, Codable, CustomStringConvertible {
    public let rawValue: String

    public init(rawValue: String) { self.rawValue = rawValue }
    public init(_ rawValue: String) { self.rawValue = rawValue }

    public init(from decoder: any Decoder) throws {
        self.init(rawValue: try decoder.singleValueContainer().decode(String.self))
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }

    public var description: String { rawValue }

    /// Another device took this identity over under the app's multi-login policy.
    public static let multiLoginPolicy = KickReason("MultiLoginPolicy")
    /// The token aged out. Recoverable: ask the host app for a new one, then reconnect.
    public static let tokenExpired = KickReason("TokenExpired")
    /// The token was revoked server-side (logout elsewhere, credential rotation).
    public static let tokenRevoked = KickReason("TokenRevoked")
    public static let userBanned = KickReason("UserBanned")
    public static let appDisabled = KickReason("AppDisabled")
    public static let quotaExceeded = KickReason("QuotaExceeded")
    /// A gateway going away for a rolling update. Emphatically *not* terminal.
    public static let serverShutdown = KickReason("ServerShutdown")
    public static let adminKick = KickReason("AdminKick")

    /// Synthesised by the SDK, never sent by the server: the HTTP upgrade itself was refused with
    /// 403, which no amount of reconnecting will change.
    public static let handshakeRejected = KickReason("HandshakeRejected")

    /// The close-reason prefix the gateway writes. See `CyaimSocketWriter.CloseAsync` server-side.
    static let closePrefix = "im-kick:"

    /// Parses a WebSocket close reason, returning `nil` when the socket died of anything else.
    public init?(closeReason: String?) {
        guard let closeReason, closeReason.hasPrefix(KickReason.closePrefix) else { return nil }
        let reason = String(closeReason.dropFirst(KickReason.closePrefix.count))
        guard !reason.isEmpty else { return nil }
        self.init(rawValue: reason)
    }

    /// Reasons that mean "do not come back": reconnecting would fail identically, forever, and a
    /// client that keeps trying is a client that hammers the gateway while showing its user
    /// nothing.
    ///
    /// `TokenExpired` is absent because it is recoverable, and `ServerShutdown` is absent because
    /// it is the single most important case to reconnect after — that is a deploy, and the node is
    /// coming back.
    public static let terminalReasons: Set<KickReason> = [
        .multiLoginPolicy,
        .tokenRevoked,
        .userBanned,
        .appDisabled,
        .adminKick,
        .handshakeRejected,
    ]

    /// True when the SDK must stop reconnecting and hand the reason to the app.
    public var isTerminal: Bool { KickReason.terminalReasons.contains(self) }
}

// MARK: - Frames

/// A request the client sends. `id` is mandatory: it is how a reply finds its caller.
struct ImRequestFrame<Body: Encodable & Sendable>: Encodable, Sendable {
    let id: String
    let target: String
    let body: Body
}

/// Body placeholder for endpoints that take no arguments. Encodes as `{}`.
public struct EmptyBody: Codable, Sendable, Hashable {
    public init() {}
}

/// A frame the server sends.
///
/// Replies and server-initiated pushes are structurally identical, which is deliberate: one decoder
/// handles both, and a push can be correlated exactly like a reply.
public struct ImFrame: Sendable, Hashable {
    /// Echoes the request id for a reply; `srv-{snowflake}` for a push.
    public let id: String

    /// Endpoint or event name, e.g. `msg.send`, `evt.message`.
    public let target: String

    /// Transport-level outcome: 0 routed, 1 endpoint threw, 2 endpoint not found.
    public let status: Int

    /// Transport-level error text.
    public let msg: String?

    public let requestTime: Int64?
    public let completeTime: Int64?

    /// Business-level result. Distinct from `status`: the transport can succeed while the endpoint
    /// refuses.
    public let body: ImBody?

    public init(
        id: String,
        target: String,
        status: Int = 0,
        msg: String? = nil,
        requestTime: Int64? = nil,
        completeTime: Int64? = nil,
        body: ImBody? = nil
    ) {
        self.id = id
        self.target = target
        self.status = status
        self.msg = msg
        self.requestTime = requestTime
        self.completeTime = completeTime
        self.body = body
    }
}

/// Business-level result, nested inside the transport frame.
public struct ImBody: Sendable, Hashable {
    public let code: ImErrorCode
    public let message: String?

    /// Server-side correlation id, present on failures and on sampled successes.
    public let traceId: String?

    /// Authoritative server clock, unix ms. Trust this over the device clock.
    public let serverTime: Int64

    public let data: JSONValue?

    public init(
        code: ImErrorCode = .ok,
        message: String? = nil,
        traceId: String? = nil,
        serverTime: Int64 = 0,
        data: JSONValue? = nil
    ) {
        self.code = code
        self.message = message
        self.traceId = traceId
        self.serverTime = serverTime
        self.data = data
    }
}

extension ImFrame: Codable {
    enum CodingKeys: String, CodingKey {
        case id, target, status, msg, requestTime, completeTime, body
    }

    /// Decoded field by field with defaults rather than by synthesis.
    ///
    /// A frame that fails to decode is a message the user never sees. A server that stops sending a
    /// field, or starts sending one this build has never heard of, must cost at most that field —
    /// never the whole frame.
    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.imString(.id) ?? ""
        target = c.imString(.target) ?? ""
        status = c.imInt(.status) ?? 0
        msg = c.imString(.msg)
        requestTime = c.imInt64(.requestTime)
        completeTime = c.imInt64(.completeTime)
        body = c.imValue(ImBody.self, .body)
    }
}

extension ImBody: Codable {
    enum CodingKeys: String, CodingKey {
        case code, message, traceId, serverTime, data
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        code = c.imInt(.code).map(ImErrorCode.init(rawValue:)) ?? .ok
        message = c.imString(.message)
        traceId = c.imString(.traceId)
        serverTime = c.imInt64(.serverTime) ?? 0
        data = c.imValue(JSONValue.self, .data)
    }
}

extension ImFrame {
    /// The payload as `T`, or `nil` when this frame has no payload or it does not fit.
    public func decodedData<T: Decodable>(as type: T.Type = T.self) -> T? {
        try? requireData(as: type)
    }

    /// The payload as `T`, throwing an `ImError` that says what went wrong instead of a
    /// `DecodingError` nobody can act on.
    public func requireData<T: Decodable>(as type: T.Type = T.self) throws -> T {
        guard let payload = body?.data, !payload.isNull else {
            throw ImError(
                code: .internalError,
                message: "\(target) returned no payload",
                traceId: body?.traceId,
                target: target
            )
        }

        do {
            return try payload.decoded(as: T.self)
        } catch {
            throw ImError(
                code: .internalError,
                message: "\(target) payload did not match \(T.self): \(error)",
                traceId: body?.traceId,
                target: target
            )
        }
    }

    /// Throws when either layer reported a failure.
    ///
    /// Both layers have to be checked and they mean different things: `status` is the transport's
    /// verdict (did the frame reach an endpoint at all), `body.code` is the endpoint's (did it
    /// agree to do the work).
    public func throwIfFailed(target requestTarget: String? = nil) throws {
        let endpoint = requestTarget ?? target

        switch status {
        case 0:
            break

        case 2:
            // No such target. Deliberately **1008 UnsupportedOperation, not 1002 NotFound**: 1002
            // means "your group does not exist", 1008 means "this deployment does not have this
            // endpoint" — which is the exact signal an SDK newer than a private-deployment server
            // produces, and the one an integrator needs verbatim rather than paraphrased.
            throw ImError(
                code: .unsupportedOperation,
                message: msg ?? "this deployment has no endpoint named \(endpoint)",
                traceId: body?.traceId,
                target: endpoint
            )

        default:
            // status == 1: the endpoint was reached and threw. The business envelope is usually
            // absent in this case, so the transport's own `msg` is all there is to report.
            throw ImError(
                code: .internalError,
                message: msg ?? "the endpoint failed",
                traceId: body?.traceId,
                target: endpoint
            )
        }

        guard let body else {
            throw ImError(code: .internalError, message: "request failed", target: endpoint)
        }

        guard body.code == .ok else {
            throw ImError(
                code: body.code,
                message: body.message ?? "request failed",
                traceId: body.traceId,
                target: endpoint
            )
        }
    }
}
